pkg/scheduler/queue/queue.go

```diff
@@ -54,229 +54,354 @@ type Request interface{}
+	lastUserIndex UserIndex
+	processed     chan nextRequestForQuerier
+	next          *availableQuerier
+}
+
+type nextRequestForQuerier struct {
+	req  Request
+	next UserIndex
+	err  error
+}
+
 func NewRequestQueue(maxOutstandingPerTenant int, forgetDelay time.Duration, queueLength *prometheus.GaugeVec, discardedRequests *prometheus.CounterVec) *RequestQueue {
 	q := &RequestQueue{
-		queues:                  newUserQueues(maxOutstandingPerTenant, forgetDelay),
+		maxOutstandingPerTenant: maxOutstandingPerTenant,
+		forgetDelay:             forgetDelay,
 		connectedQuerierWorkers: atomic.NewInt32(0),
 		queueLength:             queueLength,
 		discardedRequests:       discardedRequests,
 	}
 
-	q.cond = contextCond{Cond: sync.NewCond(&q.mtx)}
-	q.Service = services.NewTimerService(forgetCheckPeriod, nil, q.forgetDisconnectedQueriers, q.stopping).WithName("request queue")
+	q.Service = services.NewTimerService(forgetCheckPeriod, q.starting, q.forgetDisconnectedQueriers, q.stop).WithName("request queue")
 
 	return q
 }
 
-// EnqueueRequest puts the request into the queue. MaxQueries is user-specific value that specifies how many queriers can
-// this user use (zero or negative = all queriers). It is passed to each EnqueueRequest, because it can change
-// between calls.
-//
-// If request is successfully enqueued, successFn is called with the lock held, before any querier can receive the request.
-func (q *RequestQueue) EnqueueRequest(userID string, req Request, maxQueriers int, successFn func()) error {
-	q.mtx.Lock()
-	defer q.mtx.Unlock()
+func (q *RequestQueue) starting(_ context.Context) error {
+	q.stopRequested = make(chan struct{})
+	q.stopCompleted = make(chan struct{})
 
-	if q.stopped {
-		return ErrStopped
+	// These channels must not be buffered so that we can detect when dispatcherLoop() has finished.
+	q.querierOperations = make(chan querierOperation)
+	q.enqueueRequests = make(chan enqueueRequest)
+	q.availableQueriers = make(chan *availableQuerier)
+
+	go q.dispatcherLoop()
+
+	return nil
+}
+
+func (q *RequestQueue) dispatcherLoop() {
+	stopping := false
+	queues := newUserQueues(q.maxOutstandingPerTenant, q.forgetDelay)
+
+	var firstAvailableQuerier *availableQuerier
+	var lastAvailableQuerier *availableQuerier
+
+	for {
+		needToDispatchQueries := false
+
+		select {
+		case <-q.stopRequested:
+			// Nothing much to do here - fall through to the stop logic below to see if we can stop immediately.
+			stopping = true
+		case qe := <-q.querierOperations:
+			switch qe.operation {
+			case registerConnection:
+				queues.addQuerierConnection(qe.querierID)
+			case unregisterConnection:
+				queues.removeQuerierConnection(qe.querierID, time.Now())
+			case notifyShutdown:
+				queues.notifyQuerierShutdown(qe.querierID)
+			case forgetDisconnected:
+				if queues.forgetDisconnectedQueriers(time.Now()) > 0 {
+					// Removing some queriers may have caused a resharding.
+					needToDispatchQueries = true
+				}
```

Comment:
isn't there resharding in all of these cases? `removeQuerier` and `recomputeUserQueriers` are called as a result of all four operations, so tenants will likely get new queriers assigned.

Reply:
The new logic just mimics the old logic. However, your comment looks legit. I looked at the old code and I can't remember why we were not notifying the wait condition in the other cases. They're all cases where a resharding may happen, so I think we should set `needToDispatchQueries = true` in other cases too.

Reply:
Good catch, fixed in 0498225.